### ISSUE TITLE
Segmentation fault of deepin-camera on Arch

### ISSUE THREAD

**Issue by ThePirate42** [linuxdeepin/developer-center#3398]
### SUMMARY | 问题概要

I installed deepin-camera from the official repository, and when I tried to start it, the icon on the task bar flashed with no window opening. I tried to start it from terminal, and the program crashes with a segmentation fault.

### STEPS TO REPRODUCE | 复现步骤

1. try to start the program

### OBSERVED RESULT | 观察到的结果

The program crashes

### EXPECTED RESULT | 期望的结果

The program boots up correctly.

### SOFTWARE/OS VERSIONS | 软件/系统版本信息

Distribution: Arch Linux
Kernel: Linux 5.19.4.arch1-1
Desktop environment: KDE (with xorg)

deepin-camera version: 1.3.10-1

### ADDITIONAL INFORMATION | 额外补充

Terminal log:
```
read /usr/share/libimagevisualresult/filter_cube/bright.dat success,filtername:bright..
read /usr/share/libimagevisualresult/filter_cube/classic.dat success,filtername:classic..
read /usr/share/libimagevisualresult/filter_cube/print.dat success,filtername:print..
read /usr/share/libimagevisualresult/filter_cube/black.dat success,filtername:black..
read /usr/share/libimagevisualresult/filter_cube/blues.dat success,filtername:blues..
read /usr/share/libimagevisualresult/filter_cube/warm.dat success,filtername:warm..
read /usr/share/libimagevisualresult/filter_cube/gray.dat success,filtername:gray..
read /usr/share/libimagevisualresult/filter_cube/cold.dat success,filtername:cold..
read 8 CUBE/dat files...
"initFilters cost 28 ms"
2022-08-29, 00:53:23.159 [Info   ] [  ] LogFile: "/home/pietro/.cache/deepin/deepin-camera/deepin-camera.log"
2022-08-29, 00:53:23.160 [Info   ] [  ] configPath "/home/pietro/.config/deepin/deepin-camera/config.conf"
2022-08-29, 00:53:23.160 [Debug  ] [qsettingbackend.cpp  Dtk::Core::QSettingBackend::QSettingBackend 55] create config "/home/pietro/.config/deepin/deepin-camera/config.conf"
2022-08-29, 00:53:23.160 [Info   ] [  ] keys ("base.general.last_open_pic_path", "base.general.last_open_vd_path", "base.photogrid.photogrids", "base.save.picdatapath", "base.save.vddatapath", "outsetting.outformat.picformat", "outsetting.outformat.vidformat", "outsetting.resolutionsetting.resolution", "photosetting.Flashlight.Flashlight", "photosetting.audiosetting.soundswitchbtn", "photosetting.maxRecordTime.maxRecordTime", "photosetting.mirrorMode.mirrorMode", "photosetting.photosdelay.photodelays", "photosetting.photosnumber.takephotos")
/usr/lib/libavcodec.so: undefined symbol: avcodec_get_context_defaults3 V4L2_CORE: V4L2_CAP_VIDEO_CAPTURE error: Successo
V4L2_CORE: ignore the device(/dev/video1) for not have the ability to capture video.
V4L2_CORE: V4L2_CAP_VIDEO_CAPTURE error: Successo
V4L2_CORE: ignore the device(/dev/video3) for not have the ability to capture video.
deepin-camera: couldn't open /home/pietro/.config/deepin/deepin-camera/deepin-camera for read: File o directory non esistente
V4L2_CORE: ERROR opening V4L interface: File o directory non esistente
v4l2core_get_device_index dev_index(0) videodevice:5c937820 device:5c92d230
V4L2_CORE: (UVCIOC_CTRL_MAP) Error: File o directory non esistente
V4L2_CORE: (UVCIOC_CTRL_MAP) Error: File o directory non esistente
V4L2_CORE: (UVCIOC_CTRL_MAP) Error: File o directory non esistente
V4L2_CORE: (UVCIOC_CTRL_MAP) Error: File o directory non esistente
V4L2_CORE: (UVCIOC_CTRL_MAP) Error: File o directory non esistente
V4L2_CORE: (UVCIOC_CTRL_MAP) Error: File o directory non esistente
V4L2_CORE: (UVCIOC_CTRL_MAP) Error: File o directory non esistente
V4L2_CORE: (UVCIOC_CTRL_MAP) Error: File o directory non esistente
V4L2_CORE: (UVCIOC_CTRL_MAP) Error: File o directory non esistente
V4L2_CORE: requested format(34363248)  [0] -> 47504a4d
V4L2_CORE: requested format(34363248)  [1] -> 56595559
V4L2_CORE: requested format(34363248)  [2] -> 33424752
V4L2_CORE: requested format(34363248)  [3] -> 33524742
V4L2_CORE: requested format(34363248)  [4] -> 32315559
V4L2_CORE: requested format(34363248)  [5] -> 32315659
deepin-camera: no valid file extension for image file
deepin-camera: using format 1
ALSA lib pcm_dsnoop.c:566:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dmix.c:999:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
Cannot connect to server socket err = File o directory non esistente
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = File o directory non esistente
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_a52.c:1001:(_snd_pcm_a52_open) a52 is only for playback
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib pcm_dmix.c:999:(snd_pcm_dmix_open) unable to open slave
Cannot connect to server socket err = File o directory non esistente
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
V4L2_CORE: requested format(47504a4d)  [0] -> 47504a4d
V4L2_CORE: requested format(47504a4d)  [0] -> 47504a4d
Errore di segmentazione (core dump creato)
```

**Comment by BLumia** [comment]
Should be a ffmpeg4 to ffmpeg5 issue. This project seems trying to load the ffmpeg library manually but loaded the incorrect one. Not sure about the exact reason why they are using this approach...

**Comment by BLumia** [comment]
A diff patch based on the current master branch: (on https://github.com/linuxdeepin/deepin-camera/commit/6fbdeb528755f398fcc7c61fdb3c4ab66a305b98 )

<details>

``` diff
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index fa4eb70..6552e62 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -43,7 +43,7 @@ set(PROJECT_INCLUDE
     ./src/qtsingleapplication
     ./src/accessibility
     ./src/accessible
-    /usr/include
+    /usr/include/ffmpeg4.4
     /usr/include/libusb-1.0
     )
     
diff --git a/src/main.cpp b/src/main.cpp
index ef6c9e0..50c7287 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -96,9 +96,9 @@ static bool CheckFFmpegEnv()
 {
     QDir dir;
     QString path  = QLibraryInfo::location(QLibraryInfo::LibrariesPath);
-    dir.setPath(path);
+    dir.setPath(path + "/ffmpeg4.4");
     QStringList list = dir.entryList(QStringList() << (QString("libavcodec") + "*"), QDir::NoDotAndDotDot | QDir::Files);
-    if (list.contains("libavcodec.so.58")) {
+    if (list.contains("libavcodec.so")) {
         return true;
     }
     return false;
diff --git a/src/src/mainwindow.cpp b/src/src/mainwindow.cpp
index 8972250..fe699f7 100644
--- a/src/src/mainwindow.cpp
+++ b/src/src/mainwindow.cpp
@@ -795,15 +795,15 @@ void CMainWindow::slotPopupSettingsDialog()
 #endif
 }
 
-QString CMainWindow::libPath(const QString &strlib)
+QString CMainWindow::libPath(const QString &strlib, const QString &pathAppend)
 {
     QDir  dir;
     QString path  = QLibraryInfo::location(QLibraryInfo::LibrariesPath);
-    dir.setPath(path);
+    dir.setPath(path + pathAppend);
     QStringList list = dir.entryList(QStringList() << (strlib + "*"), QDir::NoDotAndDotDot | QDir::Files); //filter name with strlib
 
     if (list.contains(strlib))
-        return strlib;
+        return dir.absoluteFilePath(strlib);
 
     list.sort();
     if (list.size() > 0)
@@ -855,11 +855,11 @@ void CMainWindow::getMediaFileInfoList(const QString &path, QFileInfoList &fileL
 void CMainWindow::initDynamicLibPath()
 {
     LoadLibNames tmp;
-    QByteArray avcodec = libPath("libavcodec.so").toLatin1();
+    QByteArray avcodec = libPath("libavcodec.so", "/ffmpeg4.4").toLatin1();
     tmp.chAvcodec = avcodec.data();
-    QByteArray avformat = libPath("libavformat.so").toLatin1();
+    QByteArray avformat = libPath("libavformat.so", "/ffmpeg4.4").toLatin1();
     tmp.chAvformat = avformat.data();
-    QByteArray avutil = libPath("libavutil.so").toLatin1();
+    QByteArray avutil = libPath("libavutil.so", "/ffmpeg4.4").toLatin1();
     tmp.chAvutil = avutil.data();
     QByteArray udev = libPath("libudev.so").toLatin1();
     tmp.chUdev = udev.data();
diff --git a/src/src/mainwindow.h b/src/src/mainwindow.h
index 755942c..624a153 100644
--- a/src/src/mainwindow.h
+++ b/src/src/mainwindow.h
@@ -210,7 +210,7 @@ private:
     * @brief libPath 动态库路径
     * @param strlib 路径的字符串
     */
-    QString libPath(const QString &strlib);
+    QString libPath(const QString &strlib, const QString & pathAppend = "");
 
     /**
     * @brief getMediaFileInfoList　获得录像、图片文件列表
diff --git a/src/src/windowstatethread.cpp b/src/src/windowstatethread.cpp
index 1f370b7..3c004bd 100644
--- a/src/src/windowstatethread.cpp
+++ b/src/src/windowstatethread.cpp
@@ -23,6 +23,8 @@
 #include <DWindowManagerHelper>
 #include <QMutexLocker>
 
+#include <thread>
+
 windowStateThread::windowStateThread(QObject *parent /*= nullptr*/)
     : QThread (parent)
 {
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 7299987..3f0cb66 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -35,7 +35,7 @@ set(PROJECT_INCLUDE
     ../src/src/qtsingleapplication
     ../src/
     ./stub
-    /usr/include
+    /usr/include/ffmpeg4.4
     /usr/include/libusb-1.0
     /usr/include/SDL2
     ${PROJECT_SOURCE_DIR}/../deepin-camera

```

</details>

**Comment by ThePirate42** [comment]
What am I supposed to do?

**Comment by BLumia** [comment]
You can build it manually with the patch provided above, or wait for the package maintainer to release a patched version. If you choose the first option, feel free to let me know if you need any help while building the app :)

**Comment by BLumia** [comment]
You can build it manually with the patch provided above, or wait for the package maintainer to release a patched version. If you choose the first option, feel free to let me know if you need any help while building the app :)

**Comment by ThePirate42** [comment]
> If you choose the first option, feel free to let me know if you need any help while building the app :)

Thank you! But I don't really need it, I was just trying some random webcam apps. What I meant is: should I do a bug report to the package maintainer or something like that? Or are you going to do it/there is no need for that?

**Comment by BLumia** [comment]
> Thank you! But I don't really need it, I was just trying some random webcam apps. What I meant is: should I do a bug report to the package maintainer or something like that? Or are you going to do it/there is no need for that?

I already messaged the package maintainer, and thanks again for the report :)

**Comment by felixonmars** [comment]
Please try deepin-camera 1.3.10-2 with https://github.com/linuxdeepin/deepin-camera/pull/211 applied.

**Comment by BLumia** [comment]
Can verify fixed, tested under `1.3.10-2` and `1.4.1-1` on Arch :)
